### DISCUSSION TITLE
How to modify AST to support custom nodes

### DISCUSSION THREAD

**Discussion by SandroMaglione** [markdoc/markdoc#27] [Q&A] [answered]
Hi,

I am trying to modify the AST to add a custom element. I would like to achieve something like a `markdown-it` [plugin](https://www.npmjs.com/search?q=keywords:markdown-it-plugin), specifically adding support for `<ruby>` ([like this](https://www.npmjs.com/package/markdown-it-ruby)).

I tried to use `walk()` to navigate the AST, but then I am unsure on how to update nodes in the tree. My idea was to check every `inline` (`type="inline"`) node and split its text in two different node (`children`).

Any insight on how to achieve this? Alternatively, it would be possible to use `markdown-it` plugins?

> Note: I tried using `new Node` to create a new instance of `Node`, but currently `Node` is exported as `type`, so `new Node` doesn’t work

**Comment by mfix-stripe** [comment, marked as answer]
@SandroMaglione I believe you want to use [custom tags](https://markdoc.io/docs/tags) for your use case, creating a `ruby` tag (`{% ruby %}`) instead. 

Then, you can customize the tag's behavior in the `transform`.

**Reply by SandroMaglione** [reply]
I was aware of custom tags, but my usecase was more complex. The writer in my blog is non-technical, so he is not used to adding `{%`/`%}`. What I was trying to achieve was to avoid tags at all and just interpreting `{...|...}` as `<ruby>`.

After some work, I was able to achieve this result by adding a transform step (similar to [Table of contents](https://markdoc.io/docs/examples#table-of-contents)).

I think it would be useful for everyone if I add my process to the docs. @mfix-stripe could you guide me here? Where should I add the docs for this?

**Reply by mfix-stripe** [reply]
@SandroMaglione we want to encourage folks to use tags, or [functions](https://markdoc.io/docs/functions), for cases like you mentioned above. Stripe has dozens of "non-technical" writers using Markdoc daily, so we have seen first-hand that using `{%` is a real, but _small_, barrier to entry. We hope folks learn this syntax, because it is much more powerful that encoding custom syntax as a one-off.

**Reply by haikyuu** [reply]
Another way to go about this is to create an editor extension (provided you have control over that) that reads and writes {%%} and provides a nice GUI. A crude example would be 
```
# say we have a custom bold tag
{% bold weight="200" %}
this is bold text
{% /bold %}
```
The editor would display an input/slider along with the toggle for bold. That means more custom code, but if you have a limited amount of writers, teaching them would be easier

**Reply by mfix-stripe** [reply]
@haikyuu we hope to publish a drop-in VSCode extension and language server that will assist with this sort of thing. We have one working internally at Stripe (and it's super useful!) but we have to extract it and make it generic so others can consume it easily.

**Reply by haikyuu** [reply]
That's super great @mfix-stripe

I meant editor extension as in [lexical editor](https://lexical.dev/) extension, basically hooking into the content-editable and defining custom "renderers" for markdoc tags.

Would love to see that extension and language server soon. And I hope you make it work in the browser and framework agnostic.

**Reply by mfix-stripe** [reply]
@haikyuu depending on what you mean by "renderer", I think we have support for all the possibilities I can imagine. You can define custom `transform` functions for any of your [`Tags`](https://markdoc.io/docs/tags) or [`Nodes`](https://markdoc.io/docs/nodes) which let you control how all aspected of your content are ultimately rendered.

If you are thinking about [renderers](https://markdoc.io/docs/render) themselves (like HTML or React), they are just pure functions, so you can write your own renderer fairly straightforwardly. The [React one](https://github.com/markdoc/markdoc/blob/main/src/renderers/react/react.ts) is pretty easy to understand.

**Reply by haikyuu** [reply]
What I meant by renderer is at the contenteditable level: basically its goal is allowing a non technical user to create and customize a markdoc tag `{% tab %} ` by clicking on a ui displayed in the contenteditable instead of writing the text itself.

It's of course not in the scope of markdoc itself.

Thanks for taking the time to respond. I really love the way markdoc is extensible and versatile 💯

**Reply by mfix-stripe** [reply]
Ahh, I see @haikyuu. This is not on our current roadmap, but is a super cool idea. I could definitely see you being able to do this with some fancy component registrations. Sort of the reverse of what we have up on https://markdoc.io (<kbd>CMD+J</kbd>)

> I really love the way markdoc is extensible and versatile 💯
Thanks for the kind words!

**Comment by arranf** [comment]
@mfix-stripe Running into the same issue as above. `Node` is not exported so it's currently impossible to add to the AST. 

Use case is to wrap every heading in a separate `<section>` when rendered.

**Reply by mfix-stripe** [reply]
This is what you want: 

```js
import {Ast} from '@markdoc/markdoc'
// ...
new Ast.Node(...)
```

However, for your use case above, @arranf, I would just register a new `heading` node and just render a `<section>` in the corresponding React component (unless you aren't using React).

**Reply by arranf** [reply]
@mfix-stripe Thanks! Unfortunately I want to render a section for everything _between_ two headings (i..e a section starts with a heading and contains everything up until the end of doc or the new heading).

So I think I need to manipulate the AST! 

I'll give it a go now I can use import nodes!
